Applications/Spire/Source/Ui/DropDownList.cpp

```diff
@@ -0,0 +1,102 @@
+#include "Spire/Ui/DropDownList.hpp"
+#include <QEvent>
+#include <QHBoxLayout>
+#include "Spire/Ui/ArrayListModel.hpp"
+#include "Spire/Ui/ListItem.hpp"
+#include "Spire/Ui/ListView.hpp"
+#include "Spire/Ui/OverlayPanel.hpp"
+#include "Spire/Ui/ScrollableListBox.hpp"
+
+using namespace boost;
+using namespace Spire;
+using namespace Spire::Styles;
+
+namespace {
+  auto SCROLLABLE_LIST_STYLE(StyleSheet style) {
+    style.get(Any()).
+      set(BackgroundColor(QColor::fromRgb(0, 0, 0, 0))).
+      set(border_size(0));
+    return style;
+  }
+}
+
+DropDownList::DropDownList(ListView& list_view, QWidget* parent)
+    : QWidget(parent),
+      m_list_view(&list_view) {
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  m_scrollable_list_box = new ScrollableListBox(*m_list_view, this);
+  set_style(*m_scrollable_list_box,
+    SCROLLABLE_LIST_STYLE(get_style(*m_scrollable_list_box)));
+  layout->addWidget(m_scrollable_list_box);
+  m_panel = new OverlayPanel(this, parent);
+  m_panel->set_closed_on_blur(true);
+  on_panel_style();
+  connect_style_signal(*m_panel, [=] { on_panel_style(); });
+  parent->installEventFilter(this);
+  m_scrollable_list_box->installEventFilter(this);
+}
+
+QSize DropDownList::sizeHint() const {
+  if(!m_size_hint) {
+    auto width = std::max(m_panel->parentWidget()->size().width() -
+      m_panel_border_size.width(), m_scrollable_list_box->sizeHint().width());
+    auto height = [=] {
+      if(auto list_item = m_list_view->get_list_item(0)) {
+        return std::min(10 * list_item->sizeHint().height(),
+          m_scrollable_list_box->sizeHint().height());
+      }
+      return 0;
+    }();
+    m_size_hint.emplace(width, height);
+  }
+  return *m_size_hint;
+}
+
+bool DropDownList::eventFilter(QObject* watched, QEvent* event) {
+  if(event->type() == QEvent::Resize ||
+      event->type() == QEvent::LayoutRequest) {
+    m_size_hint = none;
+    updateGeometry();
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+bool DropDownList::event(QEvent* event) {
+  if(event->type() == QEvent::ShowToParent) {
+    if(m_list_view->get_list_model()->get_size() > 0) {
+      m_panel->show();
+      auto margins = m_panel->layout()->contentsMargins();
+      m_panel->resize(sizeHint().grownBy(margins) + m_panel_border_size);
```

Comment:
What guarantee is there that ```m_panel_border_size``` is initialized?

Reply:
on_panel_style is called in the constructor, which can initialize m_panel_border_size.